### PR TITLE
Electra hard fork

### PR DESCRIPTION
**Pull request by domiwei**
https://github.com/ethereum/consensus-specs/tree/dev/specs/electra

Skip spectest for `whisk` and `eip7594`

Mekong testnet: https://mekong.ethpandaops.io/

### CODE REVIEW

File: cl/monitor/metrics.go [hunk 1]
```diff
@@ -135,7 +136,7 @@ type TimeMeasure struct {
 }
 
 func (m TimeMeasure) End() {
-	m.metric.Set(float64(time.Since(m.start).Microseconds()))
```
Comment (Giulio2002) [inline cl/monitor/metrics.go:138]: why remotion

File: cl/phase1/core/state/accessors.go [hunk 2]
```diff
@@ -277,22 +338,25 @@ func ExpectedWithdrawals(b abstract.BeaconState, currentEpoch uint64) []*cltypes
 			}
 			withdrawals = append(withdrawals, newWithdrawal)
 			nextWithdrawalIndex++
-		} else if isPartiallyWithdrawableValidator(b.BeaconConfig(), currentValidator, currentBalance) { // Check if the validator is partially withdrawable
+		} else if isPartiallyWithdrawableValidator(b, currentValidator, currentBalance) { // Check if the validator is partially withdrawable
 			// Add a new withdrawal with the validator's withdrawal credentials and balance minus the maximum effective balance
+			amount := currentBalance - b.BeaconConfig().MaxEffectiveBalance
```
Comment (Giulio2002) [inline cl/phase1/core/state/accessors.go:343]: can't this potentially underflow?

File: cl/phase1/core/state/accessors.go [hunk 3]
```diff
@@ -277,22 +338,25 @@ func ExpectedWithdrawals(b abstract.BeaconState, currentEpoch uint64) []*cltypes
...
+			if b.Version() >= clparams.ElectraVersion {
+				amount = currentBalance - getMaxEffectiveBalanceElectra(currentValidator, b.BeaconConfig())
```
Comment (Giulio2002) [inline cl/phase1/core/state/accessors.go:345]: same as above

File: cl/phase1/core/state/cache_mutators.go [hunk 4]
```diff
@@ -77,15 +77,15 @@ func (b *CachingBeaconState) SlashValidator(slashedInd uint64, whistleblowerInd
...
+	whistleBlowerReward := newEffectiveBalance / b.BeaconConfig().GetWhistleBlowerRewardQuotient(b.Version())
 	proposerReward := b.getSlashingProposerReward(whistleBlowerReward)
 	if err := IncreaseBalance(b, proposerInd, proposerReward); err != nil {
 		return 0, err
 	}
-	rewardWhist := whistleBlowerReward - proposerReward
-	if whistleblowerInd == nil {
-		proposerReward += rewardWhist
-	}
+	//rewardWhist := whistleBlowerReward - proposerReward
```
Comment (Giulio2002) [inline cl/phase1/core/state/cache_mutators.go:85]: why not just remove?

**Review by Giulio2002 — COMMENTED** [review]
Did a partial review - another thing I would like to ask if you considered interactions with the EL too? I cannot find it. search in the specs `self.notifyNewPayload` - our equivalent are calls to `engine.NewPayload` in `forkchoice` I think that needs modifications

File: cl/monitor/metrics.go (hunk 1 above)
Comment (domiwei) [reply cl/monitor/metrics.go:138]: oh. I don't know why. thank you.

File: cl/phase1/core/state/accessors.go (hunk 2 above)
Comment (domiwei) [reply cl/phase1/core/state/accessors.go:343]: in `isPartiallyWithdrawableValidator` https://github.com/erigontech/erigon/pull/12418/files#diff-e354369205a9046bc4ccd2d5e1630441b624dfb4f4baf3ca91fb055c908e5010R119
it's guaranteed `currentBalance` must be greater than `b.BeaconConfig().MaxEffectiveBalance`

File: cl/phase1/core/state/accessors.go (hunk 3 above)
Comment (domiwei) [reply cl/phase1/core/state/accessors.go:345]: The same, `isPartiallyWithdrawableValidator` helps to check underflow is impossible

**Comment by Giulio2002** [comment]
I am going to merge this - you can do further debug on maekon

**Review by Giulio2002 — APPROVED** [review]

**Pull request merged**